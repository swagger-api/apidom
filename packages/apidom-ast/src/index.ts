// JSON AST related exports
export { default as JsonNode } from './json/nodes/JsonNode.ts';
export { default as JsonDocument } from './json/nodes/JsonDocument.ts';
export { default as JsonObject } from './json/nodes/JsonObject.ts';
export { default as JsonProperty } from './json/nodes/JsonProperty.ts';
export type { JsonValue as JsonPropertyValue } from './json/nodes/JsonProperty.ts';
export { default as JsonArray } from './json/nodes/JsonArray.ts';
export { default as JsonValue } from './json/nodes/JsonValue.ts';
export type { JsonValueOptions } from './json/nodes/JsonValue.ts';
export { default as JsonKey } from './json/nodes/JsonKey.ts';
export { default as JsonString } from './json/nodes/JsonString.ts';
export { default as JsonStringContent } from './json/nodes/JsonStringContent.ts';
export { default as JsonEscapeSequence } from './json/nodes/JsonEscapeSequence.ts';
export { default as JsonNumber } from './json/nodes/JsonNumber.ts';
export { default as JsonTrue } from './json/nodes/JsonTrue.ts';
export { default as JsonFalse } from './json/nodes/JsonFalse.ts';
export { default as JsonNull } from './json/nodes/JsonNull.ts';
export {
  isDocument as isJsonDocument,
  isFalse as isJsonFalse,
  isProperty as isJsonProperty,
  isStringContent as isJsonStringContent,
  isEscapeSequence as isJsonEscapeSequence,
  isArray as isJsonArray,
  isKey as isJsonKey,
  isNull as isJsonNull,
  isNumber as isJsonNumber,
  isObject as isJsonObject,
  isString as isJsonString,
  isTrue as isJsonTrue,
} from './json/nodes/predicates.ts';
// YAML AST related exports
export { default as YamlAlias } from './yaml/nodes/YamlAlias.ts';
export type { YamlAliasOptions } from './yaml/nodes/YamlAlias.ts';
export { default as YamlCollection } from './yaml/nodes/YamlCollection.ts';
export { default as YamlComment } from './yaml/nodes/YamlComment.ts';
export type { YamlCommentOptions } from './yaml/nodes/YamlComment.ts';
export { default as YamlDirective } from './yaml/nodes/YamlDirective.ts';
export type { YamlDirectiveOptions, YamlDirectiveParameters } from './yaml/nodes/YamlDirective.ts';
export { default as YamlDocument } from './yaml/nodes/YamlDocument.ts';
export { default as YamlKeyValuePair } from './yaml/nodes/YamlKeyValuePair.ts';
export type { YamlKeyValuePairOptions } from './yaml/nodes/YamlKeyValuePair.ts';
export { default as YamlMapping } from './yaml/nodes/YamlMapping.ts';
export { default as YamlNode } from './yaml/nodes/YamlNode.ts';
export type { YamlNodeOptions } from './yaml/nodes/YamlNode.ts';
export { default as YamlScalar } from './yaml/nodes/YamlScalar.ts';
export type { YamlScalarOptions } from './yaml/nodes/YamlScalar.ts';
export { default as YamlSequence } from './yaml/nodes/YamlSequence.ts';
export { default as YamlStream } from './yaml/nodes/YamlStream.ts';
export { default as YamlTag, YamlNodeKind } from './yaml/nodes/YamlTag.ts';
export type { YamlTagOptions } from './yaml/nodes/YamlTag.ts';
export { default as YamlAnchor } from './yaml/nodes/YamlAnchor.ts';
export type { YamlAnchorOptions } from './yaml/nodes/YamlAnchor.ts';
export { YamlStyle, YamlStyleGroup } from './yaml/nodes/YamlStyle.ts';
export { default as YamlFailsafeSchema } from './yaml/schemas/failsafe/index.ts';
export { default as YamlJsonSchema } from './yaml/schemas/json/index.ts';
export { default as YamlReferenceManager } from './yaml/anchors-aliases/ReferenceManager.ts';
export {
  isAlias as isYamlAlias,
  isAnchor as isYamlAnchor,
  isComment as isYamlComment,
  isKeyValuePair as isYamlKeyValuePair,
  isDirective as isYamlDirective,
  isDocument as isYamlDocument,
  isMapping as isYamlMapping,
  isScalar as isYamlScalar,
  isSequence as isYamlSequence,
  isStream as isYamlStream,
  isTag as isYamlTag,
} from './yaml/nodes/predicates.ts';
export { default as YamlError } from './yaml/errors/YamlError.ts';
export { default as YamlReferenceError } from './yaml/errors/YamlReferenceError.ts';
export { default as YamlSchemaError } from './yaml/errors/YamlSchemaError.ts';
export { default as YamlTagError } from './yaml/errors/YamlTagError.ts';
export type { YamlTagErrorOptions } from './yaml/errors/YamlTagError.ts';
// generic AST related exports
export type { default as Node, NodeOptions } from './Node.ts';
export { default as Literal } from './Literal.ts';
export type { LiteralOptions } from './Literal.ts';
export { Point, default as Position } from './Position.ts';
export type { PointOptions, PositionOptions } from './Position.ts';
export { default as Error } from './Error.ts';
export type { ErrorOptions } from './Error.ts';
export { default as ParseResult } from './ParseResult.ts';
export { isParseResult, isLiteral, isPoint, isPosition } from './predicates.ts';
// AST traversal related exports
export {
  BREAK,
  mergeAll as mergeAllVisitors,
  getVisitFn,
  visit,
  getNodeType,
  isNode,
  cloneNode,
} from './traversal/visitor.ts';
export type { MergeAllSync, MergeAllAsync } from './traversal/visitor.ts';
