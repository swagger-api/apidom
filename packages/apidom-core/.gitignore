/dist
/es
/cjs
/types
/NOTICE
/swagger-api-apidom-core-*.tgz
