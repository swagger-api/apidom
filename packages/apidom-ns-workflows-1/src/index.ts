export {
  isRefElement,
  isLinkElement as isLinkPrimitiveElement,
  isMemberElement,
  isObjectElement,
  isArrayElement,
  isBooleanElement,
  isNullElement,
  isElement,
  isNumberElement,
  isStringElement,
} from '@swagger-api/apidom-core';

export { default as mediaTypes, WorkflowsMediaTypes } from './media-types';

// eslint-disable-next-line no-restricted-exports
export { default } from './namespace';

export { default as refractorPluginReplaceEmptyElement } from './refractor/plugins/replace-empty-element';

export { default as refract, createRefractor } from './refractor';
export { default as specificationObj } from './refractor/specification';

export {
  isWorkflowsSpecElement,
  isWorkflowsSpecification1Element,
  isInfoElement,
  isSourceDescriptionElement,
  isSourceDescriptionsElement,
  isParameterElement,
  isSuccessActionElement,
  isSuccessActionCriteriaElement,
  isFailureActionElement,
  isFailureActionCriteriaElement,
  isComponentsElement,
  isCriterionElement,
  isReferenceElement,
  isJSONSchemaElement,
} from './predicates';

export { isWorkflowsSpecificationExtension } from './refractor/predicates';

export { default as FixedFieldsVisitor } from './refractor/visitors/generics/FixedFieldsVisitor';
export { default as MapVisitor } from './refractor/visitors/generics/MapVisitor';
export { default as MixedFieldsVisitor } from './refractor/visitors/generics/MixedFieldsVisitor';
export { default as FallbackVisitor } from './refractor/visitors/FallbackVisitor';
export { default as SpecificationVisitor } from './refractor/visitors/SpecificationVisitor';
export { default as Visitor } from './refractor/visitors/Visitor';

export { keyMap, getNodeType } from './traversal/visitor';

// Workflows 1.0.0 elements
export {
  WorkflowsSpecification1Element,
  WorkflowsSpecElement,
  InfoElement,
  SourceDescriptionElement,
  ParameterElement,
  SuccessActionElement,
  FailureActionElement,
  ComponentsElement,
  CriterionElement,
  ReferenceElement,
  JSONSchemaElement,
} from './refractor/registration';
// NCE types
export { default as SourceDescriptionsElement } from './elements/nces/SourceDescriptions';
export { default as SuccessActionCriteriaElement } from './elements/nces/SuccessActionCriteria';
export { default as FailureActionCriteriaElement } from './elements/nces/FailureActionCriteria';
export { default as ComponentsSchemas } from './elements/nces/ComponentsInputs';
export { default as ComponentsParameters } from './elements/nces/ComponentsParameters';
