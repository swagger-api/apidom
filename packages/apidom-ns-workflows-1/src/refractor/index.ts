import {
  visit,
  Element,
  dereference,
  refract as baseRefract,
  dispatchRefractorPlugins,
} from '@swagger-api/apidom-core';
import { path } from 'ramda';

import type VisitorClass from './visitors/Visitor';
import specification from './specification';
import { keyMap, getNodeType } from '../traversal/visitor';
import createToolbox from './toolbox';

const refract = <T extends Element>(
  value: unknown,
  {
    specPath = ['visitors', 'document', 'objects', 'WorkflowsSpecification', '$visitor'],
    plugins = [],
  } = {},
): T => {
  const element = baseRefract(value);
  const resolvedSpec = dereference(specification);

  /**
   * This is where generic ApiDOM becomes semantic (namespace applied).
   * We don't allow consumers to hook into this translation.
   * Though we allow consumers to define their onw plugins on already transformed ApiDOM.
   */
  const RootVisitorClass = path(specPath, resolvedSpec) as typeof VisitorClass;
  const rootVisitor = new RootVisitorClass({ specObj: resolvedSpec });

  visit(element, rootVisitor);

  /**
   * Running plugins visitors means extra single traversal === performance hit.
   */
  return dispatchRefractorPlugins(rootVisitor.element, plugins, {
    toolboxCreator: createToolbox,
    visitorOptions: { keyMap, nodeTypeGetter: getNodeType },
  }) as T;
};

export const createRefractor =
  (specPath: string[]) =>
  (value: unknown, options = {}) =>
    refract(value, { specPath, ...options });

export default refract;
