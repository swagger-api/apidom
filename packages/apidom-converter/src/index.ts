import { ParseResultElement } from '@swagger-api/apidom-core';
import { mergeOptions, bundle, File } from '@swagger-api/apidom-reference';

import defaultOptions, { ConverterOptions } from './options';
import ConvertError from './errors/ConvertError';
import UnmatchedConvertStrategyError from './errors/UnmatchedConvertStrategyError';

export { ConvertError, UnmatchedConvertStrategyError };

/**
 * `convertApiDOM` already assumes that the ApiDOM is bundled.
 */
export const convertApiDOM = async (element: ParseResultElement, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const file = File({
    uri: mergedOptions.resolve.baseURI,
    parseResult: element,
    mediaType: mergedOptions.convert.sourceMediaType || mergedOptions.parse.mediaType,
  });
  const strategy = mergedOptions.convert.strategies.find((s) => s.canConvert(file, mergedOptions));

  if (typeof strategy === 'undefined') {
    throw new UnmatchedConvertStrategyError(file.uri);
  }

  return strategy.convert(file, mergedOptions);
};

const convert = async (uri: string, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const parseResult = await bundle(uri, mergedOptions);

  return convertApiDOM(parseResult, mergedOptions);
};

export default convert;
