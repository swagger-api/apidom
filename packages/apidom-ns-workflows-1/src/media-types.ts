import { last } from 'ramda';
import { MediaTypes } from '@swagger-api/apidom-core';

type Format = 'generic' | 'json' | 'yaml';

export class WorkflowsMediaTypes extends MediaTypes<string> {
  filterByFormat(format: Format = 'generic') {
    const effectiveFormat = format === 'generic' ? 'workflows;version' : format;
    return this.filter((mediaType) => mediaType.includes(effectiveFormat));
  }

  findBy(version = '1.0.0', format: Format = 'generic') {
    const search =
      format === 'generic'
        ? `vnd.oai.workflows;version=${version}`
        : `vnd.oai.workflows+${format};version=${version}`;
    const found = this.find((mediaType) => mediaType.includes(search));

    return found || this.unknownMediaType;
  }

  latest(format: Format = 'generic') {
    return last(this.filterByFormat(format)) as string;
  }
}

const mediaTypes = new WorkflowsMediaTypes(
  'application/vnd.oai.workflows;version=1.0.0',
  'application/vnd.oai.workflows+json;version=1.0.0',
  'application/vnd.oai.workflows+yaml;version=1.0.0',
);

export default mediaTypes;
