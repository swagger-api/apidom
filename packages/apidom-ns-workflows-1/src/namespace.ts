import { NamespacePluginOptions } from '@swagger-api/apidom-core';

import WorkflowsSpecification1Element from './elements/WorkflowsSpecification1';
import WorkflowsSpecElement from './elements/WorkflowsSpec';
import InfoElement from './elements/Info';
import SourceDescriptionElement from './elements/SourceDescription';
import ParameterElement from './elements/Parameter';
import SuccessActionElement from './elements/SuccessAction';
import FailureActionElement from './elements/FailureAction';
import ComponentsElement from './elements/Components';
import CriterionElement from './elements/Criterion';
import ReferenceElement from './elements/Reference';
import JSONSchemaElement from './elements/JSONSchema';

const workflows1 = {
  namespace: (options: NamespacePluginOptions) => {
    const { base } = options;

    base.register('workflowsSpecification1', WorkflowsSpecification1Element);
    base.register('workflowsSpec', WorkflowsSpecElement);
    base.register('info', InfoElement);
    base.register('sourceDescription', SourceDescriptionElement);
    base.register('parameter', ParameterElement);
    base.register('successAction', SuccessActionElement);
    base.register('failureAction', FailureActionElement);
    base.register('components', ComponentsElement);
    base.register('criterion', CriterionElement);
    base.register('reference', ReferenceElement);
    base.register('jSONSchemaDraft202012', JSONSchemaElement);

    return base;
  },
};

export default workflows1;
