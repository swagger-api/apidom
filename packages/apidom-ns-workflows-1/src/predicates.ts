import { createPredicate } from '@swagger-api/apidom-core';

import WorkflowsSpecification1Element from './elements/WorkflowsSpecification1';
import WorkflowsSpecElement from './elements/WorkflowsSpec';
import InfoElement from './elements/Info';
import SourceDescriptionElement from './elements/SourceDescription';
import CriterionElement from './elements/Criterion';
// NCE types
import SourceDescriptionsElement from './elements/nces/SourceDescriptions';

export const isWorkflowsSpecElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: unknown): element is WorkflowsSpecElement =>
      element instanceof WorkflowsSpecElement ||
      (hasBasicElementProps(element) &&
        isElementType('workflowsSpec', element) &&
        primitiveEq('string', element));
  },
);

export const isWorkflowsSpecification1Element = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq, hasClass }) => {
    return (element: unknown): element is WorkflowsSpecification1Element =>
      element instanceof WorkflowsSpecification1Element ||
      (hasBasicElementProps(element) &&
        isElementType('workflowsSpecification1', element) &&
        primitiveEq('object', element) &&
        hasClass('api', element) &&
        hasClass('workflow', element));
  },
);

export const isInfoElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq, hasClass }) => {
    return (element: unknown): element is InfoElement =>
      element instanceof InfoElement ||
      (hasBasicElementProps(element) &&
        isElementType('info', element) &&
        primitiveEq('object', element) &&
        hasClass('info', element));
  },
);

export const isSourceDescriptionElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: unknown): element is SourceDescriptionElement =>
      element instanceof SourceDescriptionElement ||
      (hasBasicElementProps(element) &&
        isElementType('sourceDescription', element) &&
        primitiveEq('object', element));
  },
);

export const isSourceDescriptionsElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq, hasClass }) => {
    return (element: unknown): element is SourceDescriptionsElement =>
      element instanceof SourceDescriptionsElement ||
      (hasBasicElementProps(element) &&
        isElementType('sourceDescriptions', element) &&
        primitiveEq('array', element) &&
        hasClass('sourceDescriptions', element));
  },
);

export const isCriterionElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: unknown): element is CriterionElement =>
      element instanceof CriterionElement ||
      (hasBasicElementProps(element) &&
        isElementType('criterion', element) &&
        primitiveEq('object', element));
  },
);
