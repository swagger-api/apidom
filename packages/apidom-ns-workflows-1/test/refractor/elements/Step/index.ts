import { expect } from 'chai';
import { sexprs } from '@swagger-api/apidom-core';

import { StepElement } from '../../../../src';

describe('refractor', function () {
  context('elements', function () {
    context('StepElement', function () {
      specify('should refract to semantic ApiDOM tree', function () {
        const stepElement = StepElement.refract({
          description: 'Search for available pets',
          stepId: 'searchForPet',
          operationId: 'getPets',
          operationRef:
            'https://petstore3.swagger.io/api/v3/openapi.json#/paths/users/~findbystatus~1{status}/get',
          workflowId: 'uniqueWorkflowId',
          parameters: [
            {
              name: 'status',
              in: 'query',
              value: 'available',
            },
            {
              $ref: '#/json/pointer',
            },
          ],
          dependsOn: ['someStepId'],
          successCriteria: [
            {
              context: '$statusCode',
              condition: '^503$',
              type: 'regex',
            },
          ],
          onSuccess: [
            {
              type: 'goto',
              workflowId: 'uniqueWorkflowId',
              stepId: 'getPetStep',
              criteria: [
                {
                  context: '$statusCode',
                  condition: '^200$',
                  type: 'regex',
                },
              ],
            },
          ],
          onFailure: [
            {
              type: 'retry',
              workflowId: 'uniqueWorkflowId',
              stepId: 'getPetStep',
              retryAfter: 500,
              retryLimit: 5,
              criteria: [
                {
                  context: '$statusCode',
                  condition: '^503$',
                  type: 'regex',
                },
              ],
            },
          ],
          outputs: [{ key: 'value' }],
        });

        expect(sexprs(stepElement)).toMatchSnapshot();
      });
    });
  });
});
