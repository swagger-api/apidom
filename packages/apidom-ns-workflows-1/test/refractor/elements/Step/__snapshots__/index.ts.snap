// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refractor elements StepElement should refract to semantic ApiDOM tree 1`] = `
(StepElement
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (ArrayElement
      (ParameterElement
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement)))
      (ReferenceElement
        (MemberElement
          (StringElement)
          (StringElement)))))
  (MemberElement
    (StringElement)
    (ArrayElement
      (StringElement)))
  (MemberElement
    (StringElement)
    (ArrayElement
      (CriterionElement
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement)))))
  (MemberElement
    (StringElement)
    (ArrayElement
      (SuccessActionElement
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (ArrayElement
            (CriterionElement
              (MemberElement
                (StringElement)
                (StringElement))
              (MemberElement
                (StringElement)
                (StringElement))
              (MemberElement
                (StringElement)
                (StringElement))))))))
  (MemberElement
    (StringElement)
    (ArrayElement
      (FailureActionElement
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (StringElement))
        (MemberElement
          (StringElement)
          (NumberElement))
        (MemberElement
          (StringElement)
          (NumberElement))
        (MemberElement
          (StringElement)
          (ArrayElement
            (CriterionElement
              (MemberElement
                (StringElement)
                (StringElement))
              (MemberElement
                (StringElement)
                (StringElement))
              (MemberElement
                (StringElement)
                (StringElement))))))))
  (MemberElement
    (StringElement)
    (ArrayElement
      (ObjectElement
        (MemberElement
          (StringElement)
          (StringElement))))))
`;
